src/proto/factories/layer_factory.cpp

```diff
@@ -110,8 +113,196 @@ std::vector<El::Int> get_slice_points_from_reader(const generic_data_reader* dr,
                                                   const std::string& var_category,
                                                   bool& is_supported);
 
+std::string get_layer_datatype_from_pbuf(const lbann_data::Layer& proto_layer) {
+  // Get parameters from prototext
+  auto datatype_str = proto_layer.datatype();
+  if(datatype_str.empty()) {
+    datatype_str = TypeName<DataType>();
+  }
+  return datatype_str;
+}
+
+data_layout get_layer_data_layout_from_pbuf(const lbann_data::Layer& proto_layer) {
+  // Get parameters from prototext
+  const auto& layout_str = proto_layer.data_layout();
+  if (layout_str.empty())             { return data_layout::DATA_PARALLEL; }
+  if (layout_str == "data_parallel")  { return data_layout::DATA_PARALLEL; }
+  if (layout_str == "model_parallel") { return data_layout::MODEL_PARALLEL; }
+  return data_layout::invalid;
+}
+
+El::Device get_layer_device_from_pbuf(const lbann_data::Layer& proto_layer,
+                                      bool GPUs_disabled) {
+  // Get parameters from prototext
+#ifdef LBANN_HAS_GPU
+  const auto& device_str = proto_layer.device_allocation();
+  if (!GPUs_disabled) {
+    if (device_str == "gpu" || device_str.empty()) {
+      return El::Device::GPU;
+    }
+    if (device_str == "cpu") { return El::Device::CPU; }
+    if (proto_layer.has_input()) {
+      // Input layers must be on CPU
+      return El::Device::CPU;
+    }
+  }
+#endif // LBANN_HAS_GPU
+  return El::Device::CPU;
+}
+
+std::unique_ptr<Layer> build_fully_connected_layer_from_pbuf(
```

Comment:
I presume we'll eventually move this into `src/layers/learning/fully_connected.cpp`? (And similar for `build_convolution_layer_from_pbuf`?)

Reply:
Yes, we will move these after the PR.